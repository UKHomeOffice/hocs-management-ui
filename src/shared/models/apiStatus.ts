export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
