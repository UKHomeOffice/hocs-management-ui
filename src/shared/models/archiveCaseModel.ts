export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
