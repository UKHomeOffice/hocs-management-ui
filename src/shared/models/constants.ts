export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
