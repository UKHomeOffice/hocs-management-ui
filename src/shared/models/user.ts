export interface User {
    label: string;
    value: string;
}
