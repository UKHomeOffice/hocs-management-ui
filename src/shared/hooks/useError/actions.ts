import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
