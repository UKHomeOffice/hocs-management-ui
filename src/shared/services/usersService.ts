import axios from 'axios';
import { User } from '../models/user';
import Item from '../models/item';
import UpdateRequest from '../pages/user/userAmend/updateRequest';

export class AddUserError extends Error {
    userToAdd: Item;
    constructor(message: string, userToAdd: Item) {
        super(message);
        this.userToAdd = userToAdd;
    }
}

export const getUser = (userId: string) => new Promise<User>((resolve, reject) => axios
    .get(`/api/users/${userId}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getUsers = () => new Promise((resolve, reject) => axios
    .get('/api/users')
    .then((response) => {
        setTimeout(() => {
            return resolve(response.data);
        }, 2000);
    })
    .catch(reason => reject(reason))
);

export const addUserToTeam = (user: Item, teamId: string) => new Promise((resolve, reject) => axios
    .post(`/api/users/${user.value}/team/${teamId}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(new AddUserError(reason.message, user)))
);

export const deleteUserFromTeam = (userId: string, teamId: string) => new Promise((resolve, reject) => axios
    .delete(`/api/users/${userId}/team/${teamId}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const addUser = (user: FormData) => new Promise((resolve, reject) => axios
    .post('/api/users', user)
    .then((response) => resolve(response.data.userUUID))
    .catch(reason => reject(reason))
);

export const updateUser = (updateRequest: UpdateRequest) => new Promise((resolve, reject) => axios
    .put(`/api/users/${updateRequest.uuid}`, updateRequest)
    .then(() => resolve())
    .catch(reason => reject(reason))
);
