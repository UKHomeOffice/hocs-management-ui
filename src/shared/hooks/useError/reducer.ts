import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
