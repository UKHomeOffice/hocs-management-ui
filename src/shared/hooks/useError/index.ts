import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
