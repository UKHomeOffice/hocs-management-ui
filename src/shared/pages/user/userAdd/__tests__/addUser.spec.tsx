import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, fireEvent, waitFor } from '@testing-library/react';
import AddUser from '../addUser';
import * as UsersService from '../../../../services/usersService';
import { ADD_USER_BAD_REQUEST_TITLE, ADD_USER_ERROR_DESCRIPTION, ADD_USER_ERROR_TITLE } from '../../../../models/constants';
import { User } from '../../../../models/user';
import * as useError from '../../../../hooks/useError';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockUser: User;
let wrapper: RenderResult;

const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();

const userId = 'xx-xx-xx-xx';

const renderComponent = () => render(
    <MemoryRouter>
        <AddUser history={history} location={location} match={match}/>
    </MemoryRouter>
);

jest.spyOn(UsersService, 'addUser').mockImplementation(() => Promise.resolve(userId));

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: {},
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockUser = {
        id: userId,
        username: 'un',
        email: '',
        firstName: '',
        lastName: '',
        enabled: true
    };
    useReducerSpy.mockImplementation(() => [mockUser, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the addUser component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await waitFor(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeEach(async () => {
            mockUser.email = 'firstname@example.com';
            mockUser.firstName = 'firstname';
            mockUser.lastName = 'lastname';
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call is successful', () => {
            it('should redirect to the view user page', async () => {
                expect.assertions(1);

                await waitFor(() => {
                    expect(history.push).toHaveBeenCalledWith(`/user-view/${userId}`, { successMessage: 'User added successfully' });
                });
            });
            it('should clear any previous errors', async () => {
                expect.assertions(1);

                await waitFor(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails', () => {
            beforeAll(() => {
                jest.spyOn(UsersService, 'addUser').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: ADD_USER_ERROR_DESCRIPTION, title: ADD_USER_ERROR_TITLE });
            });

            it('should clear any previous errors', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });

        describe('and the service call fails with a bad request', () => {
            const errorMessage = 'errMsg';
            beforeAll(() => {
                jest.spyOn(UsersService, 'addUser').mockImplementationOnce(() => Promise.reject({ response: { status: 400, data: { message: errorMessage } } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: errorMessage, title: ADD_USER_BAD_REQUEST_TITLE });
            });

            it('should clear any previous errors', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
    });

    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'email', value: 'The Email Address is required' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(2, { key: 'firstName', value: 'The First Name is required' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(3, { key: 'lastName', value: 'The Last Name is required' });
        });
    });
    describe('and the email is not valid', () => {
        beforeEach(async () => {
            mockUser.email = 'dsfd';
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });
            fireEvent.click(submitButton);
        });

        it('should show the validation message', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'email', value: 'Email Address must be a valid email' });
        });
    });
});
