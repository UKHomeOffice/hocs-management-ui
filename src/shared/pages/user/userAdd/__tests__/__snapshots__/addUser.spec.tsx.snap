// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addUser component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add a User
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/user"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="email"
            id="email-label"
          >
            Email Address
          </label>
          <input
            class="govuk-input"
            id="email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="firstName"
            id="firstName-label"
          >
            First Name
          </label>
          <input
            class="govuk-input"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="lastName"
            id="lastName-label"
          >
            Last Name
          </label>
          <input
            class="govuk-input"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
